src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -65,7 +65,7 @@ module Crystal
     # Here we store the cumulative types of variables as we traverse the nodes.
     getter meta_vars : MetaVars
     property is_initialize : Bool
-    property exception_handler_vars : MetaVars? = nil
+    property all_exception_handler_vars : Array(MetaVars)? = nil
```

Comment:
To explain the change with the minimum amount of words: instead of tracking a single set of variables to track on exception handlers, we need a list of them to take into account nested contexts.